llmfoundry/data_prep/convert_dataset_hf.py

```diff
@@ -0,0 +1,494 @@
+        )
+
+    if concat_tokens is not None and tokenizer is not None:
+        mode = ConcatMode.CONCAT_TOKENS
+        built_tokenizer = build_tokenizer(tokenizer, tokenizer_kwargs)
+        # we will enforce length, so suppress warnings about sequences too long for the model
+        built_tokenizer.model_max_length = int(1e30)
+        columns = {'tokens': 'ndarray:int32'}
+    else:
+        mode = ConcatMode.NO_CONCAT
+        built_tokenizer = None
+        columns = {'text': 'str'}
+
+    for split_name in splits:
+        try:
+            split = dataset_constants.splits[split_name]
+        except KeyError:
+            raise KeyError(f'Constants not defined for split {split_name}.')
+        hf_split = split.hf_split
+        folder_split = split.folder_split
+        expected_num_samples = split.raw_samples
+        truncate_num_samples = split.truncated_samples
+        # Only generate the splits requested
+        if folder_split not in splits:
+            continue
+
+        # Get samples
+        hf_dataset = build_hf_dataset(
+            dataset_name=dataset,
+            data_subset=data_subset,
+            split=hf_split,
+            mode=mode,
+            max_length=concat_tokens,
+            bos_text=bos_text,
+            eos_text=eos_text,
+            no_wrap=no_wrap,
+            tokenizer=built_tokenizer,
+        )
+        loader = build_dataloader(
+            dataset=hf_dataset,
+            batch_size=512,
+            num_workers=num_workers,
+        )
+        samples = generate_samples(
+            loader,
+            truncate_num_samples=truncate_num_samples,
+        )
+
+        if expected_num_samples is not None and concat_tokens is not None:
+            denominator = truncate_num_samples if truncate_num_samples is not None else _est_progress_denominator(
+                total_samples=expected_num_samples,
+                chars_per_sample=dataset_constants.chars_per_sample,
+                chars_per_token=dataset_constants.chars_per_token,
+                mode=mode,
+                max_length=concat_tokens,
+            )
+        else:
+            denominator = None
+
+        # Write samples
+        print(f'Converting {folder_split} to MDS format...')
+        print(
+            f'Note: the progress bar is based on the dataset length before tokenization, and may finish at a value before 100%.',
+        )
+        with MDSWriter(
+            columns=columns,
+            out=os.path.join(out_root, folder_split),
+            compression=compression,
+        ) as out:
+            if denominator is not None:
+                for sample in tqdm(
+                    samples,
+                    desc=folder_split,
+                    total=denominator,
+                ):
+                    out.write(sample)
+            else:
+                for sample in tqdm(samples, desc=folder_split):
+                    out.write(sample)
+
+
+def convert_dataset_hf_from_args(
```

Comment:
I think it makes more sense to have a user call `convert_dataset_hf` which does parameter validation/parsing in a helper function as part of that function logic